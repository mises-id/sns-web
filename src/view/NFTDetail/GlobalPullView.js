/*
 * @Author: lmk
 * @Date: 2022-05-12 16:53:47
 * @LastEditTime: 2022-05-25 14:59:49
 * @LastEditors: lmk
 * @Description: 
 */
import { comment } from "@/api/comment";
import { NFTEvent, NFTLikeList } from "@/api/user";
import PullList from "@/components/PullList";
import { formatTimeStr, useChangePosts, useList, username } from "@/utils";
import React,{ useEffect, useState } from "react";
import UserHeader from "../Follows/UserHeader";
import Avatar from "@/components/NFTAvatar";

export const GlobalPullView = ({type,nftInfo})=>{
  const { followPress } = useChangePosts();
  const likeRenderView = (val,index) => {
    return <div className="nft-user-item" key={index} >
      <UserHeader
        followed={()=>followPress(val)}
        size="35px" 
        item={{...val.user,is_public:true}} />
    </div>
  }
  // comment tab content
  const commentRenderView = (val,index)=>{
    return <div>CommentContent</div>
  }
  // activity tab content
  const activeRender = (val,index) => {
    const {to_account,from_account} = val;
    const address = to_account&&to_account.address ? to_account.address.substr(0,6)+'**********' : '';
    const userInfo = {
      avatar: from_account&&from_account.mises_user ? from_account.mises_user.avatar : {
        medium:from_account&&from_account.profile_img_url ? from_account.profile_img_url : '',
        nft_asset_id:'',
      },
      username:username(to_account&&to_account.mises_user ? to_account.mises_user : {},address),
      event_type: val.event_type,
      created_at:val.created_date
    }
    return <div className="activity-user m-flex" key={index}>
      <Avatar avatarItem={userInfo.avatar} size="30px"/>
      <div className="activity-user-content">
        <div className="activity-user-name">{userInfo.event_type} To <span>{userInfo.username}</span></div>
        <div className="activity-user-time">{formatTimeStr(userInfo.created_at)}</div>
      </div>
    </div>
  }
  const typeObj = {
    like: {
      fn:NFTLikeList,
      renderView: likeRenderView
    },
    activity: {
      fn:NFTEvent,
      renderView: activeRender
    },
    comment:{
      fn:comment,
      renderView: commentRenderView
    }
  }[type]
  const [lastId, setlastId] = useState("");
  const [fetchData, last_id, dataSource] = useList(typeObj.fn, {
    id:nftInfo.id,
    nft_asset_id:nftInfo.id,
    limit: 20,
    last_id: lastId,
  });
  //getData
  useEffect(() => {
    setlastId(last_id);
  }, [last_id]);
  return <PullList
    isAuto
    key={type}
    renderView={typeObj.renderView}
    data={dataSource}
    load={fetchData}
  />
}