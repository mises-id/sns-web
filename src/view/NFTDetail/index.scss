.NFT-detail{
  &-img{
    width: 100%;
    height: calc(100vw - 30px);
    border-radius: 15px;
    overflow: hidden;
  }
  &-content{
    padding: 15px;
  }
  &-like{
    position: absolute;
    right: 0;
    width: 80px;
    height: 80px;
    bottom: -50px;
  }
  &-title{
    padding: 30px 0 20px;
    font-size: 22px;
    font-weight: bold;
    color: #333333; 
  }
  &-descitem{
    background: #FFFFFF;
    border: 1px solid #EEEEEE;
    border-radius: 8px;
    padding: 20px 0 ;
    &.grid-cols-3{
      grid-template-columns: 1fr 1fr 1fr;
    }
    .descitem{
      text-align: center;
      line-height: 1;
      width: 100%;
      &:nth-of-type(2){
        position: relative;
        &::after, &::before{
          content: "";
          width: 1px;
          height: 40px;
          background: #EEEEEE;
          position: absolute;
          top: 10px;
        }
        &::after{
          left: 0;
        }
        &::before{
          right: 0;
        }
      }
      p{
        &:first-child{
          font-size: 16px;
          font-weight: bold;
          color: #333333
        }
        &:last-child{
          margin-top: 15px;
          font-size: 13px;
          color: #666
        }
      }
    }
  }

  .link{
    font-size: 12px;
    font-weight: 400;
    color: #666666;
    justify-content: flex-end;
    padding-top: 15px;
    a{
      color: #5D61FF;
    }
    img{
      width: 4px;
      height: 7px;
      display: inline-block;
      margin-left: 4px;
    }
  }
  &-tab{
    margin-top: 10px;
    &-title{
      font-size: 13px;
      color: #666666;
    }
  }
  .NFT-detail-tab-title{
    display: block;
    padding: 9px 0;
    font-weight: bold;
  }
  .adm-tabs-tab-active{
    .NFT-detail-tab-title{
      color: var(--adm-color-primary);
    }
  }
  .tabContainer{
    position: sticky;
    top: 0;
    z-index: 79;
    background: #ffffff;
  }
  .adm-tabs-tab-line{
    height: 1px;
  }
}
.tab-container{
  .m-layout{
    height: auto;
    min-height: 400px;
    overflow: hidden;
  }
}
.nft-user-item{
  padding: 17px 15px 0;
  &:last-child{
    padding-bottom: 20px;
  }
}
.activity-user{
  padding: 25px 15px 0;
  &:last-child{
    padding-bottom: 20px;
  }
  .activity-user-content{
    margin-left: 10px;
  }
  .activity-user-name{
    font-size: 13px;
    font-weight: 400;
    color: #111111;
    span{
      font-weight: bold;
    }
  }
  .activity-user-time{
    font-size: 11px;
    font-weight: 400;
    margin-top: 9px;
    line-height: 1;
    color: #666666;
  }
}
.comment-view{
  min-height: 400px;
  padding-bottom: 60px;
}