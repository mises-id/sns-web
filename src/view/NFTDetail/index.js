/*
 * @Author: lmk
 * @Date: 2021-07-15 12:51:04
 * @LastEditTime: 2022-05-16 11:04:16
 * @LastEditors: lmk
 * @Description: NFT page
 */
import React, {useEffect,  useState } from "react";
import { useTranslation } from "react-i18next";
import "./index.scss";
import Navbar from "@/components/NavBar";
import { Image, Tabs } from "antd-mobile";
import like_NFT_big_hover from '@/images/like_NFT_big_hover.png'
import like_NFT_big from '@/images/like_NFT_big.png'
import Airdrop from '@/images/Airdrop.png'
import { useLogin } from "@/components/PostsIcons/common";
import { getNFTDetail, likeNFT,unlikeNFT } from "@/api/user";
import { useLoginModal, username, useRouteState } from "@/utils";
import { GlobalPullView } from "./GlobalPullView";
import { CommentView } from "../Post/CommentView";
import { useSelector } from "react-redux";
const NFTDetail = (props) => {
  const {t} = useTranslation();
  const state = useRouteState()
  const user = useSelector((state) => state.user) || {};
  // tab title
  const tabTitle = (title,num)=> <span className="NFT-detail-tab-title">{title}{num ? `(${num})` : ''}</span>
  const [activeKey,setActiveKey] = useState('like')
  const {isLogin} = useLogin()
  const loginModal = useLoginModal();

  const [nftInfo,setnftInfo] = useState({
    image_url:state.image,
    name:state.name,
    id:state.tokenId
  })

  useEffect(() => {
    if(nftInfo.id){
      getNFTDetail(nftInfo.id).then(res=>{
        setnftInfo(res)
      })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  // validate login
  const hasLogin = (e, fn) => {
    e.stopPropagation();
    if (!isLogin) {
      loginModal(fn, { type: "isLogin" });
      return false;
    }
    fn && fn();
  };
  // like and unlike
  const like = ()=>{
    const likeFn = nftInfo.is_liked ? unlikeNFT : likeNFT;
    likeFn(nftInfo.id).then(_=>{
      setnftInfo({
        ...nftInfo,
        is_liked: !nftInfo.is_liked
      })
    })
  }
  return (
    <div>
      <Navbar title={t("NFTDetailPageTitle")} />
      <div className="m-bg-f8f8f8 NFT-detail m-layout">
        <div className="m-bg-fff NFT-detail-content">
          <div className="m-position-relative">
            <Image src={nftInfo.image_url} className="NFT-detail-img" fit="cover"/>
            <img src={nftInfo.is_liked ? like_NFT_big_hover : like_NFT_big} alt="" onClick={e=>hasLogin(e,like)} className="NFT-detail-like"/>
          </div>
          <p className="NFT-detail-title">{nftInfo.name}</p>
          <div className="NFT-detail-descitem m-grid grid-cols-3">
            <div className="descitem">
              <p>{nftInfo.price || '0 ETH'}</p>
              <p>Floor Price</p>
            </div>
            <div className="descitem">
              <p>{nftInfo.rarity || '0%'}</p>
              <p>Rarity</p>
            </div>
            <div className="descitem">
              <p>{username(nftInfo.user)}</p>
              <p>Owner</p>
            </div>
          </div>
          <p className="link m-flex m-flex-center">
            Want to buy it?  <a href={nftInfo.perma_link} target="_blank" rel="noreferrer">Click here</a>
            <img src={Airdrop} alt="" />
          </p>
        </div>
        {/* like comment and activity tab */}
        <div className="NFT-detail-tab m-bg-fff">
          <div className="tabContainer">
            <Tabs activeLineMode="full" destroyOnClose activeKey={activeKey} onChange={setActiveKey}>
              <Tabs.Tab title={tabTitle(t('NFTDetailLike'),nftInfo.likes_count)} key='like'>
              </Tabs.Tab>
              <Tabs.Tab  title={tabTitle(t('NFTDetailComment'),nftInfo.comments_count)}  key='comment'>
              </Tabs.Tab>
              <Tabs.Tab  title={tabTitle(t('NFTDetailActivity'))}  key='activity'>
              </Tabs.Tab>
            </Tabs>
          </div>
          <div className="tab-container">
            {nftInfo.id && ['like','activity'].includes(activeKey)&&<GlobalPullView key={activeKey} type={activeKey} nftInfo={nftInfo}/> }
            {nftInfo.id && activeKey === 'comment' &&  <div className="comment-view">
              <CommentView item={{nft_asset_id:nftInfo.id,user:{uid:user.loginForm&&user.loginForm.uid}}} type="NFT"/>
            </div>}
          </div>
        </div>
      </div>
    </div>
  );
};
export default NFTDetail;
