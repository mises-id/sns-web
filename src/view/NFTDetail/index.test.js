/*
 * @Author: lmk
 * @Date: 2021-07-15 12:51:10
 * @LastEditTime: 2022-05-10 14:24:40
 * @LastEditors: lmk
 * @Description: NFTDetail
 */
import React from 'react';
import NFTDetail from '.';
import {shallow } from 'enzyme';
describe('NFTDetail', () => {
  it('should render correctly', () => {
    const component = shallow(<NFTDetail />)
    expect(component).toMatchSnapshot();
  });
});