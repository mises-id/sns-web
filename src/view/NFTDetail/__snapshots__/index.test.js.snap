// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTDetail should render correctly 1`] = `
<div>
  <Navbar
    title="NFTDetailPageTitle"
  />
  <div
    className="m-bg-f8f8f8 NFT-detail m-layout"
  >
    <div
      className="m-bg-fff NFT-detail-content"
    >
      <div
        className="m-position-relative"
      >
        <Staged
          className="NFT-detail-img"
          fit="cover"
        />
        <img
          alt=""
          className="NFT-detail-like"
          onClick={[Function]}
          src="like_NFT_big.png"
        />
      </div>
      <p
        className="NFT-detail-title"
      />
      <div
        className="NFT-detail-descitem m-grid grid-cols-3"
      >
        <div
          className="descitem"
        >
          <p>
            0 ETH
          </p>
          <p>
            Floor Price
          </p>
        </div>
        <div
          className="descitem"
        >
          <p>
            0%
          </p>
          <p>
            Rarity
          </p>
        </div>
        <div
          className="descitem"
        >
          <p>
            Anonymous
          </p>
          <p>
            Owner
          </p>
        </div>
      </div>
      <p
        className="link m-flex m-flex-center"
      >
        Want to buy it?  
        <a
          rel="noreferrer"
          target="_blank"
        >
          Click here
        </a>
        <img
          alt=""
          src="Airdrop.png"
        />
      </p>
    </div>
    <div
      className="NFT-detail-tab m-bg-fff"
    >
      <div
        className="tabContainer"
      >
        <Tabs
          activeKey="like"
          activeLineMode="full"
          destroyOnClose={true}
          onChange={[Function]}
        >
          <Tab
            key="like"
            title={
              <span
                className="NFT-detail-tab-title"
              >
                NFTDetailLike
                
              </span>
            }
          />
          <Tab
            key="comment"
            title={
              <span
                className="NFT-detail-tab-title"
              >
                NFTDetailComment
                
              </span>
            }
          />
          <Tab
            key="activity"
            title={
              <span
                className="NFT-detail-tab-title"
              >
                NFTDetailActivity
                
              </span>
            }
          />
        </Tabs>
      </div>
      <div
        className="tab-container"
      />
    </div>
  </div>
</div>
`;
